pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/table/internal/AbruptCompletionAnalysis.java

```diff
@@ -0,0 +1,335 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.symbols.table.internal;
+
+import static net.sourceforge.pmd.util.OptionalBool.NO;
+import static net.sourceforge.pmd.util.OptionalBool.UNKNOWN;
+import static net.sourceforge.pmd.util.OptionalBool.YES;
+import static net.sourceforge.pmd.util.OptionalBool.join;
+import static net.sourceforge.pmd.util.OptionalBool.max;
+import static net.sourceforge.pmd.util.OptionalBool.min;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import net.sourceforge.pmd.internal.util.AssertionUtil;
+import net.sourceforge.pmd.lang.ast.NodeStream;
+import net.sourceforge.pmd.lang.java.ast.ASTBlock;
+import net.sourceforge.pmd.lang.java.ast.ASTBreakStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTContinueStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTIfStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTLabeledStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTReturnStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTSwitchArrowBranch;
+import net.sourceforge.pmd.lang.java.ast.ASTSwitchArrowRHS;
+import net.sourceforge.pmd.lang.java.ast.ASTSwitchBranch;
+import net.sourceforge.pmd.lang.java.ast.ASTSwitchFallthroughBranch;
+import net.sourceforge.pmd.lang.java.ast.ASTSwitchStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTSynchronizedStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTThrowStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTWhileStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTYieldStatement;
+import net.sourceforge.pmd.lang.java.ast.JavaNode;
+import net.sourceforge.pmd.lang.java.rule.internal.JavaRuleUtil;
+import net.sourceforge.pmd.util.OptionalBool;
+
+/**
+ * Implementation of {@link #completesNormally(ASTStatement)}, which
+ * is used to implement scoping of pattern variables.
+ *
+ * @see <a href="https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.1">Java Language Specification</a>
+ */
+final class AbruptCompletionAnalysis {
+
+    private AbruptCompletionAnalysis() {
+        // util class
+    }
+
+    /**
+     * Returns whether the statement can complete normally. For instance,
+     * an expression statement may complete normally, while a return or throw
+     * always complete abruptly.
+     */
+    static boolean canCompleteNormally(ASTStatement stmt) {
+        return completesNormally(stmt) != NO;
+    }
+
+    static OptionalBool completesNormally(ASTStatement stmt) {
+        return completesNormally(stmt, new State(null));
+    }
+
+
+    static OptionalBool completesNormally(ASTStatement stmt, State state) {
```

Comment:
An idea: [PrematureDeclaration](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#prematuredeclaration) could be improved using this logic. It would flag variables that are declared before a statement that doesn't use it and can complete abruptly (it would recognize break and continue and yield in addition to return and throw).

It would be nice to simplify this logic though. and also for, foreach and do loops are still todo.

Reply:
In the current version of this branch, this improvement to PrematureDeclaration is no longer that easy to make. The code now implements the JLS more directly. It fully covers the JLS, and is easier to update when preview features are stabilized.